### PR TITLE
Turbowarp player in editor

### LINKED ISSUE #164: Turbowarp player in editor (player)
This replaces the existing player with turbowarp by grabbing the id from url and embedding turbowarp where the normal player would be

**Link to the new addon**

[here](https://github.com/YOYITsM3M8/ScratchAddons)

### PR DESCRIPTION
**Pull request by YOYITsM3M8**
**Resolves**

Resolves #164

**Changes**

This replaces the existing player with turbowarp by grabbing the id from url and embedding turbowarp where the normal player would be

**Reason for changes**

Turbowarp is faster than normal scratch

**Tests**

on my chromebook, chrome

@noobtracker

### CODE REVIEW

**Comment by Hans5958** [comment]
Make sure you search on the issues list first before creating another one! (#79, #150)

**Comment by NoobTracker** [comment]
> 
> 
> Make sure you search on the issues list first before creating another one! (#79, #150)

Um this is a PR.

**Comment by NoobTracker** [comment]
Ah sorry you're completely right, I thought that this is about something that replaces the stage with fork.

**Comment by YOYITsM3M8** [comment]
@NoobTracker ur right

**Comment by Hans5958** [comment]
Nevermind, you actually noticed it, but isn't it better to improve #150 than making another one?

File: addons/turbo-editor/userscript.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    var current_url = document.location; // get url
+    var id = current_url.href.split('/')[4]; // get project id from url
+    function check() { // check if page loaded
+        const el = document.querySelector('.stage-wrapper_stage-wrapper_2bejr');
+        try {
+            if (el.classList[0] == "stage-wrapper_stage-wrapper_2bejr") {
+                el.removeChild(el.childNodes[0]);
+                el.removeChild(el.childNodes[0]);
+                var turboframe = document.createElement("iframe");
+                turboframe.src="https://turbowarp.org/fullscreen.html#"+id;
```
Comment (GarboMuffin) [inline addons/turbo-editor/userscript.js:12]: instead of `https://turbowarp.org/fullscreen.html#`, you should use `https://turbowarp.org/embed.html#`

Comment (YOYITsM3M8) [reply addons/turbo-editor/userscript.js:12]: :O didn’t know that existed

Comment (NoobTracker) [reply addons/turbo-editor/userscript.js:12]: Forkphorus also has that.

**Comment by NoobTracker** [comment]
Um you should enable prettier

**Comment by NoobTracker** [comment]
just enable it here: https://github.com/YOYITsM3M8/ScratchAddons/actions

**Comment by YOYITsM3M8** [comment]
Idk how

**Comment by NoobTracker** [comment]
You just have to press a green button I think

**Comment by NoobTracker** [comment]
at least for me it was that easy

**Comment by NoobTracker** [comment]
Just click there on "actions" on your fork.

**Comment by YOYITsM3M8** [comment]
Did it work?

**Comment by NoobTracker** [comment]
yep

**Comment by NoobTracker** [comment]
Was that easy.

**Comment by YOYITsM3M8** [comment]
Oh yay

**Comment by YOYITsM3M8** [comment]
K

**Comment by NoobTracker** [comment]
So I already noticed that it also has those problems that it sometimes works and sometimes not. Somtimes there's just that null error.

**Comment by YOYITsM3M8** [comment]
idk but it works as a normal Userscript

**Comment by NoobTracker** [comment]
Same goes probably for my thing

**Comment by YOYITsM3M8** [comment]
Lol

**Comment by NoobTracker** [comment]
Anyway, just like I described it in #168, you have to do something, then save the project and then refresh the iframe with warp. That's not very helpful ....

**Comment by YOYITsM3M8** [comment]
If u enable on project page then click see inside I doesn’t work but of u reload in editor it does

**Comment by NoobTracker** [comment]
> 
> 
> If u enable on project page then click see inside I doesn’t work but of u reload in editor it does

... guess what? Same goes for my buttons.

**Comment by YOYITsM3M8** [comment]
Haha haha lol it’s bc when u click button it shows different div and it doesn’t realize it changed

**Comment by YOYITsM3M8** [comment]
@WorldLanguages could this be merged and put into beta section?

**Comment by NoobTracker** [comment]
Dunno, currently it's pretty much a mess.

**Comment by YOYITsM3M8** [comment]
:O

**Comment by WorldLanguages** [comment]
Yeah, this addon does its job but is beyond ideal. Does turbowrap make sense for the **editor**? I'd think it makes most sense for the player. While editing, we have no choice but Scratch itself.
IMO, we should have an addon that allows you to use any 3rd party player by clicking a button - and if it finds a link to turbowrap (or any other) in the project instructions/notes, it defaults to that one, instead of Scratch. If the user "sees inside", we remove the iframe and fallback to Scratch itself.

**Comment by NoobTracker** [comment]
> 
> 
> Yeah, this addon does its job but is beyond ideal. Does turbowrap make sense for the **editor**? I'd think it makes most sense for the player. While editing, we have no choice but Scratch itself.
> IMO, we should have an addon that allows you to use any 3rd party player by clicking a button - and if it finds a link to turbowrap (or any other) in the project instructions/notes, it defaults to that one, instead of Scratch. If the user "sees inside", we remove the iframe and fallback to Scratch itself.

I already started a discussion about that. #168

**Comment by YOYITsM3M8** [comment]
i was searching thru the addon.* api nad i found this: **addon.tab.getScratchVM()** could we switch out the scratch vm for [turbowarp vm](https://github.com/turbowarp/scratch-vm)? bc then instead of switching the player we could switch the vm

**Comment by NoobTracker** [comment]
That's not a bad idea - but wouldn't that just be TurboWarp? Now I had some new ideas in my PR, you may want to look at those.

**Comment by apple502j** [comment]
@YOYITsM3M8 `getScratchVM` exists but `setScratchVM` doesn't. Also, I am not still sure if it's compatible with the vanilla editor.

**Comment by NoobTracker** [comment]
I also don't know. But I think that TurboWarp itself also has ported the editor, right?

**Comment by YOYITsM3M8** [comment]
Ye they have

**Comment by NoobTracker** [comment]
so it makes not that much sense imo
since that would kinda just replace everything in the editor with TurboWarp so you can just use TurboWarp.

**Comment by apple502j** [comment]
At least remove the copyright console.log please

**Comment by NoobTracker** [comment]
> 
> 
> At least remove the copyright console.log please

bruh
It's probably not gonna be merged anyway and my PR #150 now kinda replaces this.

**Comment by YOYITsM3M8** [comment]
K

**Comment by YOYITsM3M8** [comment]
@apple502j done

**Comment by apple502j** [comment]
After reviewing and meeting with World_Languages we decided it's too similar to #150 now - closing as dupe.

**Comment by YOYITsM3M8** [comment]
Ah ok

**Pull request closed without merging**